core/src/Streamly/Internal/Unicode/Parser.hs

```diff
@@ -294,5 +297,83 @@ signed p = (negate <$> (char '-' *> p)) <|> (char '+' *> p) <|> p
 -- \"Infinity\".
 --
 -- /Unimplemented/
-double :: Parser Char m Double
-double = undefined
+-- A strict coeff and exp pair
+data CE = CE !Integer {-# UNPACK #-}!Int
+
+{-# INLINE double #-}
+double :: (Monad m) => Parser Char m Double
+double = do
+    !positive <- ((== '+') <$>  Parser.satisfy (\c -> c == '-' || c == '+')) <|>
+                    pure True
+    n <- decimal
+    let f fracDigits = CE (foldl step n fracDigits)
+                    (negate $ length fracDigits)
+        step a c = a * 10 + fromIntegral (ord c - 48)
+    CE c e <-
+        Parser.satisfy (=='.')
+            *> (f <$> Parser.takeWhile1 Char.isDigit Fold.toList)
+        <|> pure (CE n 0)
+    let !signedCoeff | positive  =  c
+                     | otherwise = -c
+    expo <-
+        Parser.satisfy (\w -> w == 'e' || w == 'E')
+            *> ((e +) <$> signed decimal)
+        <|> pure (e + 0)
+    let h sc ex = do
+            if ex >= 0
+            then fromIntegral (sc * 10 ^ ex)
+            else fromRational (sc % 10 ^ (-ex))
+    return $ h signedCoeff expo
+
+data TE =
+    TE  {-# UNPACK #-}!Int
+        {-# UNPACK #-}!Int
+        {-# UNPACK #-}!Int
+        !Integer
+        {-# UNPACK #-}!Int
+
+{-# INLINE double2 #-}
+double2 :: (Monad m) => Parser Char m Double
+double2 =  Parser step initial extract
+
+    where
+
+    initial = return $ IPartial (TE 0 0 0 0 0)
+
+    step (TE c p s m e) a = do
+        let stp = c + 1
+        return $
+            case a of
+                '-' -> Partial 0 (TE stp 0 (-1) m e)
+                '+' -> Partial 0 (TE stp 0 0 m e)
+                '.' -> Partial 0 (TE stp 1 s m e)
```

Comment:
Can these occur anywhere in the number? Can you write some common test cases to make sure this works as expected for common cases?